File: src/Filter_TEST.cc
/*
 * Copyright (C) 2012-2014 Open Source Robotics Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/

#include <gtest/gtest.h>

#include "ignition/math/Filter.hh"

using namespace ignition;

/////////////////////////////////////////////////
TEST(FilterTest, OnePole)
{
  math::OnePole<double> filterA;
  EXPECT_DOUBLE_EQ(filterA.Process(0.2), 0.0);

  filterA.Fc(0.6, 1.4);
  EXPECT_DOUBLE_EQ(filterA.Process(2.5), 2.3307710879153634);

  math::OnePole<double> filterB(0.1, 0.2);
  EXPECT_DOUBLE_EQ(filterB.Process(0.5), 0.47839304086811385);

  filterB.Set(5.4);
  EXPECT_DOUBLE_EQ(filterB.Value(), 5.4);
}

/////////////////////////////////////////////////
TEST(FilterTest, OnePoleQuaternion)
{
  math::OnePoleQuaternion filterA;
  EXPECT_EQ(filterA.Value(), math::Quaterniond(1, 0, 0, 0));

  math::OnePoleQuaternion filterB(0.4, 1.4);
  EXPECT_EQ(filterB.Value(), math::Quaterniond(1, 0, 0, 0));

  EXPECT_EQ(filterA.Process(math::Quaterniond(0.1, 0.2, 0.3)),
            math::Quaterniond(1, 0, 0, 0));

  EXPECT_EQ(filterB.Process(math::Quaterniond(0.1, 0.2, 0.3)),
            math::Quaterniond(0.98841, 0.0286272, 0.0885614, 0.119929));
}

/////////////////////////////////////////////////
TEST(FilterTest, OnePoleVector3)
{
  math::OnePoleVector3 filterA;
  EXPECT_EQ(filterA.Value(), math::Vector3d(0, 0, 0));

  math::OnePoleVector3 filterB(1.2, 3.4);
  EXPECT_EQ(filterB.Value(), math::Vector3d(0, 0, 0));

  EXPECT_EQ(filterA.Process(math::Vector3d(0.1, 0.2, 0.3)),
            math::Vector3d(0, 0, 0));

  EXPECT_EQ(filterB.Process(math::Vector3d(0.1, 0.2, 0.3)),
            math::Vector3d(0.089113, 0.178226, 0.267339));
}

/////////////////////////////////////////////////
TEST(FilterTest, Biquad)
{
  math::BiQuad<double> filterA;
  EXPECT_NEAR(filterA.Value(), 0.0, 1e-10);
  EXPECT_NEAR(filterA.Process(1.1), 0.0, 1e-10);

  filterA.Fc(0.3, 1.4);
  EXPECT_DOUBLE_EQ(filterA.Process(1.2), 0.66924691484768517);

  filterA.Fc(0.3, 1.4, 0.1);
  EXPECT_DOUBLE_EQ(filterA.Process(10.25), 0.96057152402651302);

  math::BiQuad<double> filterB(4.3, 10.6);
  EXPECT_NEAR(filterB.Value(), 0.0, 1e-10);
  EXPECT_DOUBLE_EQ(filterB.Process(0.1234),  0.072418159950486546);

  filterB.Set(4.5);
  EXPECT_DOUBLE_EQ(filterB.Value(), 4.5);
}

/////////////////////////////////////////////////
TEST(FilterTest, BiquadVector3)
{
  math::BiQuadVector3 filterA;
  EXPECT_EQ(filterA.Value(), math::Vector3d(0, 0, 0));
  EXPECT_EQ(filterA.Process(math::Vector3d(1.1, 2.3, 3.4)),
            math::Vector3d(0, 0, 0));

  math::BiQuadVector3 filterB(6.5, 22.4);
  EXPECT_EQ(filterB.Value(), math::Vector3d(0, 0, 0));
  EXPECT_EQ(filterB.Process(math::Vector3d(0.1, 20.3, 33.45)),
            math::Vector3d(0.031748, 6.44475, 10.6196));
}

File: src/python/Filter_TEST.py
# Copyright (C) 2021 Open Source Robotics Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License")
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest
from ignition.math import BiQuadd
from ignition.math import BiQuadVector3
from ignition.math import OnePoled
from ignition.math import OnePoleQuaternion
from ignition.math import OnePoleVector3
from ignition.math import Quaterniond
from ignition.math import Vector3d


class TestFilter(unittest.TestCase):

    def test_one_pole(self):
        filter_a = OnePoled()
        self.assertAlmostEqual(filter_a.process(0.2), 0.0)

        filter_a.fc(0.6, 1.4)
        self.assertAlmostEqual(filter_a.process(2.5), 2.3307710879153634)

        filter_b = OnePoled(0.1, 0.2)
        self.assertAlmostEqual(filter_b.process(0.5), 0.47839304086811385)

        filter_b.set(5.4)
        self.assertAlmostEqual(filter_b.value(), 5.4)

    def test_one_pole_quaternion(self):
        filter_a = OnePoleQuaternion()
        self.assertAlmostEqual(filter_a.value(), Quaterniond(1, 0, 0, 0))

        filter_b = OnePoleQuaternion(0.4, 1.4)
        self.assertAlmostEqual(filter_b.value(), Quaterniond(1, 0, 0, 0))

        self.assertAlmostEqual(filter_a.process(Quaterniond(0.1, 0.2, 0.3)),
                               Quaterniond(1, 0, 0, 0))

        self.assertAlmostEqual(filter_b.process(Quaterniond(0.1, 0.2, 0.3)),
                               Quaterniond(0.98841, 0.0286272,
                                           0.0885614, 0.119929))

    def test_one_pole_vector3(self):
        filter_a = OnePoleVector3()
        self.assertAlmostEqual(filter_a.value(), Vector3d(0, 0, 0))

        filter_b = OnePoleVector3(1.2, 3.4)
        self.assertAlmostEqual(filter_b.value(), Vector3d(0, 0, 0))

        self.assertAlmostEqual(filter_a.process(Vector3d(0.1, 0.2, 0.3)),
                               Vector3d(0, 0, 0))

        self.assertAlmostEqual(filter_b.process(Vector3d(0.1, 0.2, 0.3)),
                               Vector3d(0.089113, 0.178226, 0.267339))

    def test_biquad(self):
        filter_a = BiQuadd()
        self.assertAlmostEqual(filter_a.value(), 0.0, delta=1e-10)
        self.assertAlmostEqual(filter_a.process(1.1), 0.0, delta=1e-10)

        filter_a.fc(0.3, 1.4)
        self.assertAlmostEqual(filter_a.process(1.2), 0.66924691484768517)

        filter_a.fc(0.3, 1.4, 0.1)
        self.assertAlmostEqual(filter_a.process(10.25), 0.96057152402651302)

        filter_b = BiQuadd(4.3, 10.6)
        self.assertAlmostEqual(filter_b.value(), 0.0, delta=1e-10)
        self.assertAlmostEqual(filter_b.process(0.1234),  0.072418159950486546)

        filter_b.set(4.5)
        self.assertAlmostEqual(filter_b.value(), 4.5)

    def test_biquad_vector3(self):
        filter_a = BiQuadVector3()
        self.assertEqual(filter_a.value(), Vector3d(0, 0, 0))
        self.assertEqual(filter_a.process(Vector3d(1.1, 2.3, 3.4)),
                         Vector3d(0, 0, 0))

        filter_b = BiQuadVector3(6.5, 22.4)
        self.assertEqual(filter_b.value(), Vector3d(0, 0, 0))
        self.assertEqual(filter_b.process(Vector3d(0.1, 20.3, 33.45)),
                         Vector3d(0.031748, 6.44475, 10.6196))


if __name__ == '__main__':
    unittest.main()
